### PR TITLE
[Westminster] Add OpenID Connect single sign-on support

### PR DESCRIPTION
**Pull request by davea**
Extends social login functionality to authenticate against an (MS Active Directory B2C-specific) OpenID Connect identity provider.

### CODE REVIEW

File: perllib/FixMyStreet/DB/Result/User.pm [hunk 1]
```diff
@@ -54,16 +62,10 @@ __PACKAGE__->add_columns(
...
-  "twitter_id",
-  { data_type => "bigint", is_nullable => 1 },
-  "facebook_id",
-  { data_type => "bigint", is_nullable => 1 },
-  "extra",
-  { data_type => "text", is_nullable => 1 },
   "area_ids",
   { data_type => "integer[]", is_nullable => 1 },
+  "oidc_ids",
+  { data_type => "text[]", is_nullable => 1 },
```
Comment (dracos) [inline perllib/FixMyStreet/DB/Result/User.pm]: Not that it matters much, but if you generate this on a fresh db schema created from scratch, it wouldn't move the other rows around.

File: perllib/FixMyStreet/DB/Result/User.pm [hunk 2]
```diff
@@ -654,4 +658,20 @@ sub in_role {
...
+    my @oidc_ids = uniq ( $oidc_id, @$oidc_ids );
+    $self->oidc_ids(\@oidc_ids);
+}
+
+sub remove_oidc_id {
+    my ($self, $oidc_id) = @_;
+
+    my $oidc_ids = $self->oidc_ids || [];
+    my @oidc_ids = grep { $_ ne $oidc_id } @$oidc_ids;
+    $self->oidc_ids(\@oidc_ids);
```
Comment (dracos) [inline perllib/FixMyStreet/DB/Result/User.pm]: To match a new user (as first test), should this set to undef if empty?

File: perllib/FixMyStreet/App/Controller/Auth.pm [hunk 3]
```diff
@@ -187,6 +188,8 @@ sub email_sign_in : Private {
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{facebook_id};
     $token_data->{twitter_id} = $c->session->{oauth}{twitter_id}
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{twitter_id};
+    $token_data->{oidc_id} = $c->session->{oauth}{oidc_id}
+        if $c->get_param('oauth_need_email') && $c->session->{oauth}{oidc_id};
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Auth.pm]: Feels like these three could be in a little loop now, given they're all the same.

File: perllib/FixMyStreet/App/Controller/Report/New.pm [hunk 4]
```diff
@@ -1191,6 +1191,8 @@ sub tokenize_user : Private {
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{facebook_id};
     $c->stash->{token_data}{twitter_id} = $c->session->{oauth}{twitter_id}
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{twitter_id};
+    $c->stash->{token_data}{oidc_id} = $c->session->{oauth}{oidc_id}
+        if $c->get_param('oauth_need_email') && $c->session->{oauth}{oidc_id};
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Report/New.pm]: Ditto

File: perllib/FixMyStreet/App/Controller/Auth.pm [hunk 5]
```diff
@@ -47,6 +47,7 @@ sub general : Path : Args(0) {
     if (!$data_username && !$data_password && !$data_email) {
         $c->detach('social/facebook_sign_in') if $c->get_param('facebook_sign_in');
         $c->detach('social/twitter_sign_in') if $c->get_param('twitter_sign_in');
+        $c->detach('social/oidc_sign_in') if $c->get_param('oidc_sign_in');
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Auth.pm]: Should we rename this `sign_in` with three different values - looks like that would help simplify this in a few places in this commit? Or is there a reason they're separate.

File: perllib/FixMyStreet/App/Controller/Auth/Social.pm [hunk 6]
```diff
@@ -142,6 +143,84 @@ sub twitter_callback: Path('/auth/Twitter') : Args(0) {
...
+
+    my %oauth;
+    $oauth{return_url} = $c->get_param('r');
+    $oauth{detach_to} = $c->stash->{detach_to};
+    $oauth{detach_args} = $c->stash->{detach_args};
+    $c->session->{oauth} = \%oauth;
+    $c->res->redirect($url);
+}
+
+sub oidc_callback: Path('/auth/OIDC') : Args(0) {
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Auth/Social.pm]: I realise they're all subtly different, but is there any worth in factoring the (now) three out to shared subclasses or something with variants for their differing parts?

File: perllib/FixMyStreet/App/Controller/Report/Update.pm [hunk 7]
```diff
@@ -408,6 +408,8 @@ sub tokenize_user : Private {
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{facebook_id};
     $c->stash->{token_data}{twitter_id} = $c->session->{oauth}{twitter_id}
         if $c->get_param('oauth_need_email') && $c->session->{oauth}{twitter_id};
+    $c->stash->{token_data}{oidc_id} = $c->session->{oauth}{oidc_id}
+        if $c->get_param('oauth_need_email') && $c->session->{oauth}{oidc_id};
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Report/Update.pm]: Ditto the first thing. Or a shared fn or something.

File: perllib/FixMyStreet/App/Controller/Report/Update.pm [hunk 8]
```diff
@@ -444,6 +446,8 @@ sub save_update : Private {
             $c->detach('/auth/social/facebook_sign_in');
         } elsif ( $c->get_param('twitter_sign_in') ) {
             $c->detach('/auth/social/twitter_sign_in');
+        } elsif ( $c->get_param('oidc_sign_in') ) {
+            $c->detach('/auth/social/oidc_sign_in');
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Report/Update.pm]: Ditto ditto the second thing, unless I've missed something.

File: perllib/OIDC/Lite/Client/WebServer/Azure.pm [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+=cut
+
+
+sub get_id_token {
+    my $self = shift;
+
+    my $old_parser = $self->{response_parser};
+
+    $self->{response_parser} = OIDC::Lite::Client::IDTokenResponseParser->new;
```
Comment (dracos) [inline perllib/OIDC/Lite/Client/WebServer/Azure.pm]: This looks fine, but is there any reason you don't set the parser once in a `new` function; it looks like it's only refresh_access_token is other fn that uses it and presumably that's not called.

Comment (davea) [reply perllib/OIDC/Lite/Client/WebServer/Azure.pm]: That's a better idea, thanks.

File: perllib/FixMyStreet/DB/Result/User.pm (hunk 1 above)
Comment (davea) [reply perllib/FixMyStreet/DB/Result/User.pm]: Good to know for next time 👍

File: perllib/FixMyStreet/App/Controller/Auth.pm (hunk 5 above)
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Auth.pm]: Do you mean the function name or the parameter?

Comment (dracos) [reply perllib/FixMyStreet/App/Controller/Auth.pm]: I think I meant parameter here.
But I think I mentioned function elsewhere, though probably more work/less worth it.

File: perllib/FixMyStreet/App/Controller/Auth.pm [hunk 10]
```diff
@@ -44,10 +44,8 @@ sub general : Path : Args(0) {
 
     # decide which action to take
     $c->detach('code_sign_in') if $clicked_sign_in_by_code || ($data_email && !$data_password);
-    if (!$data_username && !$data_password && !$data_email) {
-        $c->detach('social/facebook_sign_in') if $c->get_param('facebook_sign_in');
-        $c->detach('social/twitter_sign_in') if $c->get_param('twitter_sign_in');
-        $c->detach('social/oidc_sign_in') if $c->get_param('oidc_sign_in');
+    if (!$data_username && !$data_password && !$data_email && $c->get_param('social_sign_in')) {
+        $c->detach('social/handle_sign_in');
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Auth.pm]: I think all the calls to `social/handle_sign_in` should be forward, not detach, that way the current behaviour of carrying on if bad data provided is maintained, I think.

**Pull request merged**